test/controller/mixins/spec.check-progress.js

```diff
@@ -63,71 +52,295 @@ describe('mixins/check-progress', () => {
+            controller.decodeConditions.returns({
+                url: 'nextstep',
+                condition: { continueOnEdit: true }
+            });
             controller.setStepComplete(req, res);
-            req.sessionModel.get('steps').should.deep.equal([
-                '/one',
-                '/three',
-                '/two'
+            controller.addJourneyHistoryStep.should.have.been.calledWithExactly(
+                req,
+                res,
+                {
+                    path: '/base/teststep',
+                    next: '/base/nextstep',
+                    fields: undefined,
+                    skip: undefined,
+                    continueOnEdit: true
+                }
+            );
+        });
+
+    });
+
+    describe('addJourneyHistoryStep', () => {
+        it('creates a step history and add step if there is no step history', () => {
+            controller.addJourneyHistoryStep(req, res,
+                { path: '/path/newstep', next: '/path/newnext', newitem: true }
+            );
+            req.journeyModel.get('history').should.deep.equal([
+                { path: '/path/newstep', next: '/path/newnext', newitem: true }
+            ]);
+        });
+
+        it('adds step to end of history', () => {
+            req.journeyModel.set('history', [
+                { path: '/path/one', next: '/path/two' },
+                { path: '/path/two', next: '/path/three' }
+            ]);
+            controller.addJourneyHistoryStep(req, res,
+                { path: '/path/newstep', next: '/path/newnext', newitem: true }
+            );
+            req.journeyModel.get('history').should.deep.equal([
+                { path: '/path/one', next: '/path/two' },
+                { path: '/path/two', next: '/path/three' },
+                { path: '/path/newstep', next: '/path/newnext', newitem: true }
+            ]);
+        });
+
+        it('overwrites existing step with the same path', () => {
+            req.journeyModel.set('history', [
+                { path: '/path/one', next: '/path/two' },
+                { path: '/path/two', next: '/path/three' },
+                { path: '/path/three', next: '/path/four' },
+                { path: '/path/four', next: '/path/five' },
+                { path: '/path/five', next: '/path/six' }
+            ]);
+            controller.addJourneyHistoryStep(req, res,
+                { path: '/path/three', next: '/path/four', newitem: true }
+            );
+            req.journeyModel.get('history').should.deep.equal([
+                { path: '/path/one', next: '/path/two' },
+                { path: '/path/two', next: '/path/three' },
+                { path: '/path/three', next: '/path/four', newitem: true },
+                { path: '/path/four', next: '/path/five' },
+                { path: '/path/five', next: '/path/six' }
+            ]);
+        });
+
+        it('overwrites existing step and truncates if the next url is different', () => {
```

Comment:
If step 4/5 answers were still valid, would you handle the route change by branching? Trying to think of a specific example case.